File: tools/regmsg_shell.py
#!/usr/bin/env python3
"""
Regmsg Shell - Interactive command line interface for regmsgd

This script provides an interactive shell to communicate with the regmsgd daemon
using ZeroMQ. It allows sending commands and receiving responses in a user-friendly format.

Usage:
    python regmsg_shell.py
"""

import zmq
import binascii
import sys
import time
import argparse
from datetime import datetime

try:
    import readline  # For command history (Unix/Mac)

    HAS_READLINE = True
except ImportError:
    HAS_READLINE = False

# Default ZeroMQ endpoint for regmsgd
ENDPOINT_DEFAULT = "ipc:///var/run/regmsgd.sock"
# Default timeout for requests in milliseconds
TIMEOUT = 5000


class RegmsgShell:
    """
    An interactive shell for communicating with regmsgd daemon

    This class handles the connection to the regmsgd daemon and provides
    an interactive command line interface for sending commands and receiving responses.
    """

    def __init__(self, endpoint=ENDPOINT_DEFAULT, timeout=TIMEOUT):
        """
        Initialize the RegmsgShell with connection parameters

        Args:
            endpoint (str): ZeroMQ endpoint to connect to
            timeout (int): Request timeout in milliseconds
        """
        self.endpoint = endpoint
        self.timeout = timeout
        self.ctx = None
        self.sock = None
        self.command_history = []

    def connect(self):
        """
        Connect to the regmsgd daemon with error handling

        Returns:
            bool: True if connection was successful, False otherwise
        """
        try:
            # Create ZeroMQ context and socket
            self.ctx = zmq.Context()
            self.sock = self.ctx.socket(zmq.REQ)

            # Set socket timeouts to prevent hanging
            self.sock.setsockopt(zmq.RCVTIMEO, self.timeout)
            self.sock.setsockopt(zmq.SNDTIMEO, self.timeout)

            # Connect to the endpoint
            self.sock.connect(self.endpoint)

            # Test the connection by sending a simple command
            self.sock.send_string("listCommands")
            reply = self.sock.recv()
            print("✅ Connected to regmsgd")
            return True

        except zmq.Again:
            # This happens when the request times out
            print(f"❌ Timeout connecting to {self.endpoint}. Is the daemon running?")
            return False
        except Exception as e:
            # Handle other connection errors
            print(f"❌ Error connecting to regmsgd: {e}")
            return False

    def format_response(self, reply, start_time=None):
        """
        Format and display the response from the daemon

        Args:
            reply (bytes): Raw response from the daemon
            start_time (float, optional): Timestamp when request was sent for timing info
        """
        if start_time:
            elapsed = time.time() - start_time
            print(f"⏱️  Response time: {elapsed:.3f}s")

        try:
            # Try to decode as UTF-8 text
            text = reply.decode("utf-8")
            print("📥 Response (text):")
            print(text)
        except UnicodeDecodeError:
            # If text decoding fails, display as hexadecimal
            print("📥 Response (hex):")
            print(binascii.hexlify(reply).decode())
        except Exception as e:
            # Handle other parsing errors
            print(f"📥 Error parsing response: {e}")

    def run(self):
        """
        Run the main interactive command loop
        """
        # This check ensures self.sock is not None, satisfying static analysis
        if self.sock is None:
            print("❌ Socket is not connected. Cannot run shell.")
            return

        print("Enter a command or 'exit' to quit.")
        print("Use 'listCommands' to see available commands.")
        print("-" * 80)

        while True:
            try:
                import socket

                # Show hostname in prompt
                prompt = f"🎮 [{socket.gethostname()}]> "
                cmd = input(prompt).strip()

                if not cmd:
                    continue
                if cmd.lower() in ("exit", "quit", "q"):
                    break

                # Add to command history if readline is available
                if HAS_READLINE and cmd not in self.command_history:
                    self.command_history.append(cmd)

                # Record start time to measure response time
                start_time = time.time()

                # Send command to the daemon
                try:
                    self.sock.send_string(cmd)
                except zmq.Again:
                    print("❌ Timeout sending command")
                    continue
                except Exception as e:
                    print(f"❌ Error sending command: {e}")
                    break

                # Receive response from the daemon
                try:
                    reply = self.sock.recv()
                    self.format_response(reply, start_time)
                except zmq.Again:
                    print("❌ Timeout waiting for response")
                except Exception as e:
                    print(f"❌ Error receiving response: {e}")
                    break

                print("-" * 80)

            except KeyboardInterrupt:
                # Handle Ctrl+C gracefully
                print("\nUse 'exit' or 'quit' to exit.")
            except EOFError:
                # Handle Ctrl+D (or end of input)
                print("\nExiting...")
                break

    def close(self):
        """
        Close the connection to the daemon
        """
        if self.sock:
            self.sock.close()
        if self.ctx:
            self.ctx.term()


def main():
    """
    Main function to run the regmsg shell with argument parsing
    """
    parser = argparse.ArgumentParser(description="Regmsg Shell - CLI for regmsgd")
    parser.add_argument(
        "--endpoint",
        default=ENDPOINT_DEFAULT,
        help="ZeroMQ endpoint (default: %(default)s)",
    )
    parser.add_argument(
        "--timeout",
        type=int,
        default=TIMEOUT,
        help="Request timeout in milliseconds (default: %(default)s)",
    )

    args = parser.parse_args()

    # Create and initialize the shell
    shell = RegmsgShell(endpoint=args.endpoint, timeout=args.timeout)

    if not shell.connect():
        sys.exit(1)

    try:
        # Run the interactive shell
        shell.run()
    finally:
        # Always close the connection
        shell.close()


if __name__ == "__main__":
    main()

File: src/bin/shell/regmsg_shell.rs
//! Regmsg Shell - Interactive command line interface for regmsgd
//!
//! This binary provides an interactive shell to communicate with the regmsgd daemon
//! using ZeroMQ. It allows sending commands and receiving responses in a user-friendly format.

use std::io::{self, Write};
use std::time::Instant;

use clap::Parser;
use zmq;

/// Default ZeroMQ endpoint for regmsgd
const ENDPOINT_DEFAULT: &str = "ipc:///var/run/regmsgd.sock";
/// Default timeout for requests in milliseconds
const TIMEOUT: i32 = 5000;

/// Regmsg Shell - CLI for regmsgd
#[derive(Parser, Debug)]
#[command(author, version, about, long_about = None)]
struct Args {
    /// ZeroMQ endpoint
    #[arg(long, default_value = ENDPOINT_DEFAULT)]
    endpoint: String,

    /// Request timeout in milliseconds
    #[arg(long, default_value_t = TIMEOUT)]
    timeout: i32,
}

struct RegmsgShell {
    #[allow(dead_code)] // Kept for connection reuse and settings
    endpoint: String,
    #[allow(dead_code)] // Kept for connection reuse and settings
    timeout: i32,
    #[allow(dead_code)] // Context needs to be kept alive for the connection
    context: zmq::Context,
    socket: zmq::Socket,
}

impl RegmsgShell {
    /// Initialize the RegmsgShell with connection parameters
    fn new(endpoint: String, timeout: i32) -> Result<Self, Box<dyn std::error::Error>> {
        let context = zmq::Context::new();
        let socket = context.socket(zmq::REQ)?;

        // Set socket timeouts to prevent hanging
        socket.set_rcvtimeo(timeout)?;
        socket.set_sndtimeo(timeout)?;

        // Connect to the endpoint
        socket.connect(&endpoint)?;

        Ok(RegmsgShell {
            endpoint,
            timeout,
            context,
            socket,
        })
    }

    /// Connect to the regmsgd daemon with error handling
    fn connect(&mut self) -> Result<bool, Box<dyn std::error::Error>> {
        // Test the connection by sending a simple command
        if let Err(e) = self.socket.send("listCommands", 0) {
            eprintln!("🔴 Connection failed: {}", e);
            eprintln!(
                "💡 Make sure regmsgd daemon is running and accessible at {}",
                self.endpoint
            );
            return Ok(false);
        }

        match self.socket.recv_string(0) {
            Ok(_) => Ok(true),
            Err(e) => {
                eprintln!("🔴 Connection failed: {}", e);
                eprintln!(
                    "💡 Make sure regmsgd daemon is running and accessible at {}",
                    self.endpoint
                );
                Ok(false)
            }
        }
    }

    /// Format and display the response from the daemon
    fn format_response(&self, reply: &[u8], start_time: Option<Instant>) {
        if let Some(start_time) = start_time {
            let elapsed = start_time.elapsed();
            println!("⏱️  Response time: {:.3}s", elapsed.as_secs_f64());
        }

        // Decode as UTF-8 text, replace invalid sequences with
        let text = String::from_utf8_lossy(reply);

        // Check if response indicates an error
        if text.starts_with("Error:") || text.starts_with("Err:") {
            eprintln!("{}", text);
        } else {
            // Try to parse as JSON and pretty-print if possible
            if let Ok(json_value) = serde_json::from_str::<serde_json::Value>(&text) {
                println!(
                    "{}",
                    serde_json::to_string_pretty(&json_value).unwrap_or_else(|_| text.to_string())
                );
            } else {
                // If not JSON, print as-is (without the "🟢 Response:" header)
                println!("{}", text);
            }
        }
    }

    /// Show help information
    fn show_help(&self) {
        let help_text = r#"
🎮 Regmsg Shell Help
====================
Shell Commands:
  help            - Show this help message
  clear           - Clear the screen
  exit/quit/q     - Exit the shell

Daemon Commands:
  listCommands    - Show all available commands from regmsgd
  [other commands] - All other commands are sent to the regmsgd daemon
"#;
        println!("{}", help_text);
    }

    /// Handle internal shell commands
    fn handle_internal_command(&mut self, cmd: &str) -> Result<bool, Box<dyn std::error::Error>> {
        let cmd_lower = cmd.trim().to_lowercase();

        match cmd_lower.as_str() {
            "exit" | "quit" | "q" => {
                return Ok(true); // Signal to exit
            }
            "help" | "h" | "?" => {
                self.show_help();
                return Ok(true); // Command handled, don't send to daemon
            }
            "clear" => {
                // Clear screen using ANSI escape codes
                print!("\x1B[2J\x1B[1;1H");
                return Ok(true); // Command handled, don't send to daemon
            }
            _ => {
                // Not an internal command, continue with normal processing
            }
        }

        Ok(false)
    }

    /// Run the main interactive command loop
    fn run(&mut self) -> Result<(), Box<dyn std::error::Error>> {
        println!("Use 'help' for shell commands.");
        println!("Enter a command or 'exit' to quit.");
        println!("💡 Daemon commands: Use 'listCommands' to see available commands from regmsgd.");
        println!("{}", "─".repeat(80));

        loop {
            // Show hostname in prompt
            let hostname = get_hostname();
            let prompt = format!("[{}]> ", hostname);

            // Print the prompt and flush to ensure it appears immediately
            print!("{}", prompt);
            io::stdout().flush()?;

            // Read a line of input from the user
            let mut input = String::new();
            match io::stdin().read_line(&mut input) {
                Ok(_) => {}
                Err(e) => {
                    eprintln!("\n🔴 Error reading input: {}", e);
                    continue;
                }
            }

            // Remove trailing newline
            let cmd = input.trim();

            if cmd.is_empty() {
                continue;
            }

            // Check if it's an internal command
            match self.handle_internal_command(cmd)? {
                true => {
                    // Command was handled internally (exit, help, clear, etc.)
                    // Check if it was an exit command
                    let cmd_lower = cmd.trim().to_lowercase();
                    if cmd_lower == "exit" || cmd_lower == "quit" || cmd_lower == "q" {
                        break; // Exit the shell
                    }
                    // For other internal commands (help, clear), just continue to next iteration
                    continue;
                }
                false => {
                    // Command was not handled internally, send to daemon
                }
            }

            // Record start time to measure response time
            let start_time = Some(Instant::now());

            // Send command to the daemon
            match self.socket.send(cmd, 0) {
                Ok(_) => {}
                Err(e) => {
                    eprintln!("🔴 Error sending command: {}", e);
                    eprintln!("💡 Check your connection or command syntax");
                    continue;
                }
            }

            // Receive response from the daemon
            match self.socket.recv_bytes(0) {
                Ok(reply) => {
                    self.format_response(&reply, start_time);
                }
                Err(e) => {
                    eprintln!("🔴 Error receiving response: {}", e);
                    eprintln!(
                        "💡 The command may have timed out or the daemon may be unresponsive"
                    );
                }
            }

            println!("{}", "-".repeat(80));
        }

        Ok(())
    }
}

/// Get the hostname of the current machine
fn get_hostname() -> String {
    match hostname::get() {
        Ok(name) => name.to_string_lossy().to_string(),
        Err(_) => "localhost".to_string(),
    }
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let args = Args::parse();

    let mut shell = RegmsgShell::new(args.endpoint, args.timeout)?;

    if !shell.connect()? {
        eprintln!("🔴 Failed to connect to regmsgd. The daemon may not be running.");
        std::process::exit(1);
    }

    shell.run()?;

    Ok(())
}
